cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -52,57 +54,283 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
+					nil,
+					"--tls-bootstrap-token must also specificy --masters",
+				),
+			)
+		}
+		if len(cfg.DiscoveryToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--tls-bootstrap-token must also specificy --discovery-token",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryFileArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryFile) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryURL) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file is mutually exclusive with --token, --discovery-token, and --discovery-url",
+				),
+			)
+		}
+		if len(cfg.Masters) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file must not specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file must also specificy --tls-bootstrap-token",
+				),
+			)
+		}
+	}
 	return allErrs
 }
 
-func ValidateHTTPSDiscovery(c *kubeadm.HTTPSDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateDiscoveryURLArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryURL) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url is mutually exclusive with --token, --discovery-token, and --discovery-file",
+				),
+			)
+		}
+		if len(cfg.Masters) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must not specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must also specificy --tls-bootstrap-token",
+				),
+			)
+		}
+	}
 	return allErrs
 }
 
-func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateTLSBootstrapToken(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if len(c.ID) == 0 || len(c.Secret) == 0 {
+	id, secret, err := tokenutil.ParseToken(c.TLSBootstrapToken)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
+	}
+
+	if len(id) == 0 || len(secret) == 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath, nil, "token must be specific as <ID>:<Secret>"))
 	}
-	if len(c.Addresses) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, nil, "at least one address is required"))
+
+	return allErrs
+}
+
+func ValidateJoinMastersArgs(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
```

Comment:
should use https://golang.org/pkg/net/#SplitHostPort for validation here if masters is set